src/components/DraggableList/DraggableList.jsx

```diff
@@ -0,0 +1,226 @@
+		`,
+		dragIndex: number`
+			Index of the item the drag was started on
+		`,
+		dragOverIndex: number`
+			Index of the item the dragged item is hovered over
+		`,
+		onDragStart: func`
+			Called when the user starts to drag an item.
+			Signature: \`(dragIndex, { event, props }) => {}\`
+		`,
+		onDragEnd: func`
+			Called when the user stops to dragging an item.
+			Signature: \`({ event, props }) => {}\`
+		`,
+		onDragOver: func`
+			Called when the user drags an item over another item.
+			Signature: \`(dragOverIndex, { event, props }) => {}\`
+		`,
+		onDrop: func`
+			Called when the user drops an item in the list
+			Signature: \`({oldIndex, newIndex}, { event, props }) => {}\`
+		`,
+	},
+
+	getDefaultProps() {
+		return {
+			hasDragHandle: true,
+			onDragStart: _.noop,
+			onDragEnd: _.noop,
+			onDragOver: _.noop,
+			onDrop: _.noop,
+		};
+	},
+
+	isValidDropIndex(index) {
+		const { dragIndex } = this.props;
+		return index < dragIndex || index > dragIndex + 1;
+	},
+
+	handleDragStart(index) {
+		return event => {
+			const { dataTransfer } = event;
+			dataTransfer.effectAllowed = 'move';
+			dataTransfer.dropEffect = 'move';
+			dataTransfer.setData('drag', 'drag');
+			this.props.onDragStart(index, { event, props: this.props });
+		};
+	},
+
+	handleDragEnd(event) {
+		const { dragIndex, dragOverIndex, onDragEnd, onDrop } = this.props;
+		onDragEnd({ event, props: this.props });
+		if (this.isValidDropIndex(dragOverIndex)) {
+			onDrop(
+				{
+					oldIndex: dragIndex,
+					newIndex:
+						dragOverIndex > dragIndex ? dragOverIndex - 1 : dragOverIndex,
+				},
+				{ event, props: this.props }
+			);
+		}
+	},
+
+	handleDragOver(index) {
+		return event => {
+			event.preventDefault();
+			const { dragOverIndex, onDragOver } = this.props;
+			if (dragOverIndex !== index) {
+				onDragOver(index, { event, props: this.props });
+			}
+		};
+	},
+
+	handleDragLeave(event) {
+		const { dragIndex, onDragOver } = this.props;
+		const childCount = findTypes(this.props, DraggableList.Item).length;
+
+		if (this.lastItemEl) {
+			const { bottom } = this.lastItemEl.getBoundingClientRect();
+			if (_.isFinite(dragIndex) && event.clientY > bottom) {
+				onDragOver(childCount, { event, props: this.props });
+			}
+		}
+	},
+
+	render() {
+		const {
+			style,
+			className,
+			hasDragHandle,
+			dragIndex,
+			dragOverIndex,
+			...passThroughs
+		} = this.props;
+
+		const itemChildProps = _.map(
+			findTypes(this.props, DraggableList.Item),
+			'props'
+		);
+
+		const dividerIndex =
+			_.isNumber(dragIndex) && this.isValidDropIndex(dragOverIndex)
+				? dragOverIndex
+				: -1;
+
+		return (
+			<div
+				{...omitProps(passThroughs, DraggableList)}
+				className={cx(
+					'&',
+					{
+						'&-is-dragging': _.isNumber(dragIndex),
+					},
+					className
+				)}
+				style={style}
+				onDragLeave={this.handleDragLeave}
+			>
+				{_.map(itemChildProps, (itemChildProp, index) => {
+					return (
+						<div key={index}>
```

Comment:
I haven't thought too hard about it, but could using the index as the `key` cause some mis-renders in this component if the order of items changes in the source data?

Reply:
That's a really good question.  I hate using indexes for keys at all, but it gets a little tricky when you have unknown children.  I see using indexes as keys a lot in this code base, but this is kind of a special case because you will actively be reordering the children.  

Do we have any other patterns for handling this type of situation?  I could maybe append whatever `getCombinedChildText` returns for each item?  Not sure if that would cause performance issues though.